datafusion/functions-array/src/rewrite.rs

```diff
@@ -0,0 +1,211 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+//! Rewrites for using Array Functions
+
+use crate::concat::{array_append, array_concat};
+use crate::expr_fn::{array_has_all, array_prepend};
+use datafusion_common::config::ConfigOptions;
+use datafusion_common::tree_node::Transformed;
+use datafusion_common::utils::list_ndims;
+use datafusion_common::{Column, DFSchema};
+use datafusion_expr::expr::ScalarFunction;
+use datafusion_expr::expr_rewriter::FunctionRewrite;
+use datafusion_expr::{
+    BinaryExpr, BuiltinScalarFunction, Expr, GetFieldAccess, GetIndexedField, Operator,
+};
+
+/// Rewrites expressions into function calls to array functions
+pub(crate) struct ArrayFunctionRewriter {}
+
+impl FunctionRewrite for ArrayFunctionRewriter {
+    fn name(&self) -> &str {
+        "FunctionRewrite"
+    }
+
+    fn rewrite(
+        &self,
+        expr: Expr,
+        schema: &DFSchema,
+        _config: &ConfigOptions,
+    ) -> datafusion_common::Result<Transformed<Expr>> {
+        let transformed = match expr {
```

Comment:
The whole point of this PR is to move these rules out of `datafusion/optimizer/src/analyzer/rewrite_expr.rs`  in the optimizer crate and into the functions-array crate

Reply:
After reading this PR, a question for me is in https://github.com/apache/arrow-datafusion/pull/9563, `get_field` function also works for a single `ScalarValue` of `Struct` or `Map`
Should I put the rewrite logic in this file or create a new file under "datafusion/functions"? 🤔

Reply:
I suggest we add another match in this same file

We could also add another FunctionRewriter but that would result in another (unecessary walk down the tree I think)

Reply:
Yeah, make sense. I thought it's OK for function but may not suitable for semantic, but now adding in the same file works, we may separate another file if there're more function_rewrite.